eth2/utils/merkle_proof/src/lib.rs

```diff
@@ -1,6 +1,138 @@
+#[macro_use]
+extern crate lazy_static;
+
 use ethereum_types::H256;
 use hashing::hash;
 
+const MAX_TREE_DEPTH: usize = 32;
+const EMPTY_SLICE: &[H256] = &[];
+
+lazy_static! {
+    /// Cached zero hashes where `ZERO_HASHES[i]` is the hash of a Merkle tree with 2^i zero leaves.
+    static ref ZERO_HASHES: Vec<H256> = {
+        let mut hashes = vec![H256::from([0; 32]); MAX_TREE_DEPTH + 1];
+
+        for i in 0..MAX_TREE_DEPTH {
+            hashes[i + 1] = hash_concat(hashes[i], hashes[i]);
+        }
+
+        hashes
+    };
+
+    /// Zero nodes to act as "synthetic" left and right subtrees of other zero nodes.
+    static ref ZERO_NODES: Vec<MerkleTree> = {
+        (0..MAX_TREE_DEPTH + 1).map(MerkleTree::Zero).collect()
+    };
+}
+
+/// Right-sparse Merkle tree.
+///
+/// Efficiently represents a Merkle tree of fixed depth where only the first N
+/// indices are populated by non-zero leaves (perfect for the deposit contract tree).
+#[derive(Debug)]
+pub enum MerkleTree {
```

Comment:
I'm noticing that this is the third merkle tree implementation we're considering using in Lighthouse:

1. `cached_tree_hash`: Implemented using a contiguous byte array and an [overlay](https://github.com/sigp/lighthouse/blob/9a356a00c2c166c6c31c5de42487821c8098b791/eth2/utils/cached_tree_hash/src/btree_overlay.rs#L66-L232).
2. #417: SSZ partials -- not yet implemented.
3. This one

Perhaps there's some gain in unifying them?

Also, out of interest, does this tree have the same memory downsides mentioned here: https://doc.rust-lang.org/src/alloc/collections/btree/map.rs.html#29-35

Reply:
Yeah definitely agree about trying to unify them, I've picked that commit out of this PR and preserved it on the `merkle-tree` branch of my fork for later use.

My `MerkleTree` implementation is just a textbook binary tree implementation, and therefore it does suffer from the cache and allocation downsides mentioned in the link you posted. I figured it would be OK for testing, but for prod use, I'll take the time to understand your BTree-inspired overlay.